src/Integration/Binding/ProjectBindingOperation.Writer.cs

```diff
@@ -15,57 +15,48 @@
         {
             get;
-        } = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+        } = new Dictionary<string, RuleSet>(StringComparer.OrdinalIgnoreCase);
 
-        public ProjectRuleSetWriter()
-            :this(null)
-        {
-        }
-
-        internal /*for testing purposes*/ ProjectRuleSetWriter(IRuleSetGenerationFileSystem fs)
-            : base(fs)
-        {
-        }
 
         /// <summary>
-        /// Write a project-level SonarQube <see cref="RuleSet"/> file in the <param name="projectRoot"/> directory.
+        /// Pends a write to a project-level SonarQube <see cref="RuleSet"/> file in the <param name="projectRoot"/> directory.
         /// </summary>
         /// <param name="projectFullPath">The absolute full path to the project</param>
         /// <param name="ruleSetFileName">The rule set file name</param>
         /// <param name="solutionRuleSetPath">Full path of the parent solution-level SonarQube rule set</param>
         /// <param name="existingRuleSetPath">Existing project rule set</param>
         /// <returns>Full file path of the file that was written out</returns>
-        public string WriteProjectLevelRuleSet(string projectFullPath, string ruleSetFileName, string solutionRuleSetPath, string currentRuleSetPath)
+        internal /*for testing purposes*/ string PendWriteProjectLevelRuleSet(string projectFullPath, string ruleSetFileName, string solutionRuleSetPath, string currentRuleSetPath)
         {
-            if (string.IsNullOrWhiteSpace(projectFullPath))
-            {
-                throw new ArgumentNullException(nameof(projectFullPath));
-            }
-
-            if (string.IsNullOrWhiteSpace(ruleSetFileName))
-            {
-                throw new ArgumentNullException(nameof(ruleSetFileName));
-            }
-
-            if (string.IsNullOrWhiteSpace(solutionRuleSetPath))
-            {
-                throw new ArgumentNullException(nameof(solutionRuleSetPath));
-            }
+            Debug.Assert(!string.IsNullOrWhiteSpace(projectFullPath));
+            Debug.Assert(!string.IsNullOrWhiteSpace(ruleSetFileName));
+            Debug.Assert(!string.IsNullOrWhiteSpace(solutionRuleSetPath));
 
             string projectRoot = Path.GetDirectoryName(projectFullPath);
             string ruleSetRoot = PathHelper.ForceDirectoryEnding(projectRoot);
 
             string existingRuleSetPath;
-            if (this.TryUpdateExistingProjectRuleSet(solutionRuleSetPath, ruleSetRoot, currentRuleSetPath, out existingRuleSetPath))
+            RuleSet existingRuleSet;
+            if (this.TryUpdateExistingProjectRuleSet(solutionRuleSetPath, ruleSetRoot, currentRuleSetPath, out existingRuleSetPath, out existingRuleSet))
             {
                 Debug.Assert(existingRuleSetPath != null);
+                Debug.Assert(existingRuleSet != null);
+
+                // Pend update
```

Comment:
Now we will pend an update or a new file instead of actually modifying them. The code was refactored so that method will be the one that calls to sourceControlledFileSystem.PendFileWrite instead of this being done in multiple places.